API/ev3_robot.c

```diff
@@ -0,0 +1,163 @@
+/*
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/
+ *
+ * The Initial Developer of this code is Simón Rodriguez Perez.
+ * Portions created by Simón Rodriguez Perez are Copyright (C) 2014-2015 Simón Rodriguez Perez.
+ * All Rights Reserved.
+ *
+ */
+#include "ev3_robot.h"
+
+
+typedef struct {
+	uint8_t Pos_x;
+	uint8_t Pos_y;
+	short Head;
+} ROBOT_POSE;
+
+typedef struct {
+	ROBOT_POSE Pose;
+	int Width;
+	float WheelDiameter;
+	float WheelCircunference;
+	int ColorLeft;
+	int ColorRight;
+	uint8_t MotorLeft;
+	uint8_t MotorRight;
+	uint8_t MotorDual;
+	int Gyro;
+	float Kfriction; // friction factor if robot glide in the competition table surface
+} RobotGlobals;
+
+RobotGlobals MyRobot;
+
+typedef struct {
+	float Kp;
+	float Ki;
+	float Kd;
+} KPID;
+
+KPID StraightPid;
+
+void PoseInit(){
+	MyRobot.Pose.Pos_x = 0;
+	MyRobot.Pose.Pos_y = 0;
+	MyRobot.Pose.Head = 0;
+}
+
+void SetKPID(float Kp, float Ki, float Kd){
+
+	StraightPid.Kp = 2.0;
+	StraightPid.Ki = 0.0;
+	StraightPid.Kd = 0.0;
+	if (Kp != 0) StraightPid.Kp = Kp;	
+	if (Ki != 0) StraightPid.Ki = Ki;
+	if (Kd != 0) StraightPid.Kd = Kd;
+
+}
+
+void RobotInit(int ColorLeft, int ColorRight, uint8_t MotorLeft, uint8_t MotorRight, int Gyro, bool Debug){
+	
+	PoseInit();
+	SetKPID(0,0,0);
+
+	MyRobot.WheelCircunference = MyRobot.WheelDiameter * M_PI;
+	MyRobot.ColorLeft = ColorLeft;
+	MyRobot.ColorRight = ColorRight;
+	MyRobot.MotorLeft = MotorLeft;
+	MyRobot.MotorRight = MotorRight;
+	MyRobot.MotorDual = MotorLeft + MotorRight;
+	MyRobot.Gyro = Gyro;
+	MyRobot.Kfriction = 1.0;
+	
+	int reflectedLightLeft = ReadEV3ColorSensorReflectedLight(MyRobot.ColorLeft);
+	int reflectedLightRight = ReadEV3ColorSensorReflectedLight(MyRobot.ColorRight);
+	
+	ResetEV3GyroSensor(MyRobot.Gyro, EV3GyroHardwareCommand0x88); // chack Gyro Veersion //
+	ResetEV3GyroSensor(MyRobot.Gyro, EV3GyroHardwareReboot); // 2.5s, valid for all versions
+	ResetEV3GyroSensor(MyRobot.Gyro, EV3GyroSoftwareOffset); /* chack for proper reset */
```

Comment:
We have doubt, about this. In EV3-G we have a block to make the calibration and and entry menu to do it. 
Always previous at first mission in FLL competition, we make a Gyro Reset. It is a headhache when the gyro sensor goes crazy, and there's no way of knowing when it's going to go faulty
Thanks for your explanation!

Reply:
Indeed, the true recalibration should theoretically be done only once at the program start, subsequent resets should be `EV3GyroSoftwareOffset`. In practise, it may be the case that the angle does start to drift away after a while. In that case, either a new recalibration is needed, or one needs to do the rotation rate -> angle integration themselves and do own rate offset detection (idle sensor wouldn't report 0 °/s).

Reply:
we tried a experiment in EV3-G, and the gyro sensor when robot moves is not equal to 0 deg/s, it has a lot of noise, it may be that we are doing something wrong. The problem is where the sensor is drifting, in the middle of a straight move with gyrosensor. Maybe a solution is read the rotation rate and if it is over a value, reset the gyro sensor, but unfortunately is impossible to return to the 0 position, because you must stop the motors to reset the gyro, and the gyro is drifting in the meanwhile. We try that there are no sudden movements that seem more problematic.
We take this note, if it seems necesary to make a reset in other part of the program  we will use EV3GyroSoftwareOffset,

Reply:
>The problem is where the sensor is drifting, in the middle of a straight move with gyrosensor

If you stop the robot, does the rate return to 0 deg/s? If yes, then the problem might on the physics side of things (see hypothesis below), but if the rate is still nonzero, then the gyro is the problem. Also, is the drift happening all the time during the robot travel, or is there a single time point at which it starts drifting?

This seems like something I may have observed on the leJOS experiment mentioned previously. IIRC my robot was lightly steering to the right instead of going straight.

I have a hypothesis about why this is happening, but I haven't analyzed or tested it. The core of the problem may be that the lego gyro behaves like a single-axis gyroscope (ignoring the TILT-ANG/TILT-RATE modes). This has interesting consequences - one can turn the gyro three times along its non-sensing axes such that the gyro is rotated 90° to the original position; yet the reported sensor angle should stay unchanged. If something similar on much smaller scale happens due to vibration generated when the robot is moving, the accumulated angle may get slowly skewed towards one direction.